### PR TITLE
Feat/dns verification

### PR DESCRIPTION
**Pull request by sumitnitsurat**
- DNS verification added for document store
- Added unit test for dns verification 
- Create New issue for tls library used in dnsprove

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,6 +53,7 @@
     "redux": "^4.0.0",
     "redux-devtools-extension": "^2.13.7",
     "redux-saga": "^0.16.0",
+    "tls": "0.0.1",
```
Comment (rjchow) [inline package.json]: what's this library used for?

File: package.json [hunk 2]
```diff
@@ -40,6 +40,7 @@
     "next-seo": "^1.11.2",
     "node-sass": "^4.12.0",
     "number-to-words": "^1.2.4",
+    "opencerts-dnsprove": "^2.0.0",
```
Comment (yehjxraymond) [inline package.json]: Needs to be in @govtechsg org.

File: package.json (hunk 1 above)
Comment (yehjxraymond) [reply package.json]: If it's suppose to be in dnsprove, shoudl move there instead.

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js [hunk 3]
```diff
@@ -56,7 +57,25 @@ const renderIcon = status => {
   );
 };
 
-const renderText = status => {
+export const getIdentityVerificationText = identityStatus => {
+  const verifiedText = identityStatus.reduce(
+    (prev, next) => {
+      /* eslint-disable-next-line no-param-reassign */
+      if (next.registry && !prev.isRegistry) prev.isRegistry = true;
```
Comment (Nebulis) [inline src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: minor comment but the condition `!prev.isRegistry` is not necessary

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    const testValue = [{ registry: "Govtech", dns: "abc.com" }];
+    expect(getIdentityVerificationText(testValue)).toEqual("Accredited by SSG");
+  });
+
+  test("when registry is unverified but dns is verified", () => {
+    const testValue = [{ registry: false, dns: "abc.com" }];
+    expect(getIdentityVerificationText(testValue)).toEqual("Issued by ABC.COM");
+  });
+
+  test("when multiple registry is verified", () => {
```
Comment (Nebulis) [inline src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: let's differentiate the test case in their description

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  });
+  test("when multiple dns is verified", () => {
+    const testValue = [
+      { registry: false, dns: "abc.com" },
+      { registry: false, dns: "demo.com" }
+    ];
+    expect(getIdentityVerificationText(testValue)).toEqual(
+      "Issued by ABC.COM & DEMO.COM"
+    );
+  });
```
Comment (Nebulis) [inline src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: What about those use cases:
- testValue = []
- testValue = [{ registry: "demo", dns: "demo.com" }, { registry: false, dns: "abc.com" }]
- testValue = [{ registry: false, dns: "abc.com" }, { registry: "demo", dns: "demo.com" }]

Comment (Nebulis) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: and so ?it's still possible to call that function that way

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js (hunk 4 above)
Comment (Nebulis) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: test description must reflect what the test is ding, there are the same description twice

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js (hunk 3 above)
Comment (Nebulis) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: up to the opener to resolve discussion :)

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js (hunk 4 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: something like this "should return appropriate display text when multiple registry is verified"

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js (hunk 3 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: okay :p

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js (hunk 5 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: when there is no registry and dns saga test will fail. this component code will never execute.

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js (hunk 4 above)
Comment (Nebulis) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: yup this sounds better

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js (hunk 3 above)
Comment (Nebulis) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: What's the issue on setting isRegistry again and again ?

Anyway I actually have another suggestion to change the function:

```
if(some(identityStatus, (idontknowhowtonamethis) => idontknowhowtonamethis.isRegistry)) {
  return "Accredited by SSG";
} else {
// note filter Boolean is to remove empty values
  return `Issued by ${identityStatus.map((idontknowhowtonamethis) =>idontknowhowtonamethis.dns).filter(Boolean).dns.join(" & ")}`
}
```

the issue with the current function is that there is `no comments/explanation` and the intent using a reducer is not really clear as it's doing 2 things at the same time

**Review by rjchow — CHANGES_REQUESTED** [review]
the verification logic looks correct but the flow is hard to follow - we'll need to refactor this in the following days.

aside from that there's also the rendering issue which i posted a screenshot of on telegram

do add in the integration test cases below as well for the various combinations

### Single Issuer
##### Valid (Green)
######Schema 1.5:
Accredited Issuer

###### Schema 2.0
Accredited Issuer
DNS Issuer

##### Invalid (means it should show red and not progress)
###### Schema 1.5
Issuer not on registry
###### Schema 2.0
No identityProof, issuer not on registry
Has identityProof, DNS records has wrong address
Has identityProof, no OpenAtts records

### Multiple Issuer
##### Valid
All accredited issuers: should show accredited cert
All DNS issuers: should show issued by DNS names
Mix of accredited issuers & DNS issuers: should show accredited cert

##### Invalid (show red and not render cert)
Mix of valid and invalid issuers
All invalid issuers

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js (hunk 3 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: We are already iterating on the same array twice in our code ..with this method in case of dns it will iterate 4 times on the same array to return us the required value..

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js (hunk 4 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: updated please check .

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js (hunk 3 above)
Comment (Nebulis) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: What's the issue with that ? performance concern to iterate over tiny array ?

readabiliy > performance unless performance is an actual issue. On top of that this function iterate only twice in that case

Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: ok

File: src/sagas/certificate.js [hunk 6]
```diff
@@ -253,94 +247,144 @@ function isApprovedENSDomain(issuerAddress) {
...
-    const contractStoreAddresses = get(data, "issuers", []).map(issuer =>
-      getDocumentStore(issuer)
+export function* verifyCertificateDnsIssuer({ issuer }) {
+  const location = get(issuer, "identityProof.location");
+
+  if (!location) return false;
+
+  const dnsRecords = yield call(getDocumentStoreRecords, location);
+
+  trace(`DNS records: ${dnsRecords}`); // dnsRecords: [{addr: "0xabc", netId: 3}]
```
[the remark this answers is not in the document]
Comment (sumitnitsurat) [reply src/sagas/certificate.js]: fixed it.

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js (hunk 5 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js]: Should I resolve this ?

File: src/components/CertificateVerifyBlock/CertificateVerifyBlock.js (hunk 3 above)
Comment (sumitnitsurat) [reply src/components/CertificateVerifyBlock/CertificateVerifyBlock.js]: fixed it ..please check

File: src/sagas/certificate.test.js [hunk 7]
```diff
@@ -77,37 +94,42 @@ function whenThereIsOnlyOneEnsName() {
...
+function whenThereAreDnsNameAndEthereumAddress() {
+  const dnsName = "abc.com";
+  const ethereumAddress = "0x0c9d5E6C766030cc6f0f49951D275Ad0701F81E2";
   const testCert = new MakeCertUtil()
-    .addIssuer(ensNames[0])
-    .addIssuer(ensNames[1])
+    .addDataField("issuers", [
+      {
+        identityProof: { type: "DNS-TXT", location: dnsName },
+        documentSotre: ethereumAddress
```
[the remark this answers is not in the document]
Comment (sumitnitsurat) [reply src/sagas/certificate.test.js]: fixed it.

**Comment by sumitnitsurat** [comment]
> the verification logic looks correct but the flow is hard to follow - we'll need to refactor this in the following days.
> 
> aside from that there's also the rendering issue which i posted a screenshot of on telegram
> 
> do add in the integration test cases below as well for the various combinations
> 
> ### Single Issuer
> ##### Valid (Green)
> ######Schema 1.5:
> Accredited Issuer
> 
> ###### Schema 2.0
> Accredited Issuer
> DNS Issuer
> 
> ##### Invalid (means it should show red and not progress)
> ###### Schema 1.5
> Issuer not on registry
> 
> ###### Schema 2.0
> No identityProof, issuer not on registry
> Has identityProof, DNS records has wrong address
> Has identityProof, no OpenAtts records
> 
> ### Multiple Issuer
> ##### Valid
> All accredited issuers: should show accredited cert
> All DNS issuers: should show issued by DNS names
> Mix of accredited issuers & DNS issuers: should show accredited cert
> 
> ##### Invalid (show red and not render cert)
> Mix of valid and invalid issuers
> All invalid issuers

I will create a new issue for that to refactor and add integration test for this.

**Comment by rjchow** [comment]
do the tests in this merge so that we can have confidence that the verification logic works - right now we don't run any tests to check that cert verification is correct so that's very dangerous

**Review by rjchow — COMMENTED** [review]
left two comments the rest are correct

**Review by rjchow — APPROVED** [review]
